IconThemePlugin.js

```diff
@@ -0,0 +1,26 @@
+class IconThemePlugin {
+  apply(compiler) {
+    const pluginName = IconThemePlugin.name;
+    const { webpack } = compiler;
+    const { RawSource } = webpack.sources;
+
+    compiler.hooks.emit.tap(pluginName, (compilation) => {
+      const icons = compilation.getAssets().map(
+        (asset) => asset.name.split('/').pop()
+      );
+
+      const iconEntries = icons.map((file) => [
+        file.substr(0, file.lastIndexOf(".")),
+        file.substr(file.lastIndexOf(".") + 1, file.length)
+      ]);
+      const iconTypes = Object.fromEntries(iconEntries);
+
+      const metadata = require('./metadata.json');
+      metadata.icons = iconTypes;
+      const json = JSON.stringify(metadata, null, 2);
+      compilation.emitAsset('../metadata.json', new RawSource(json));
```

Comment:
I'm not entirely sure it's completely safe to hard-code this one. An edge-case I can think of is that someone changes their dist to be something deep.

As mentioned in an earlier thread, maybe it's a way to detect the root directory ?

Reply:
I'm not seeing a good way to detect the project root from within a Webpack plugin; how about if I add class options that give the user the ability to override the path in a parameter if they choose?

I *think* that's what you were suggesting in the next comment, but I am not sure.

Reply:
> I'm not seeing a good way to detect the project root from within a Webpack plugin;

Hm. I vaguely remember being able to grab the path to the physical `webpack.config.<ext>` file from within a plugin, then just do a dirname on it to extract the root. Even if that memory is false, I would imagine that there is a function or variable somewhere to get that.

> I think that's what you were suggesting in the next comment, but I am not sure.

Yes.

Reply:
The functions I saw to accomplish that all required changes to the actual `webpack.config.js` to enable it, but I will keep looking to make sure I haven't missed something obvious. If I can't find anything I'll push the constructor with options and we can go from there.

Reply:
I found a weird way to do it, since I didn't see any easy officially documented methods. I used `compilation.compiler.context` to get the project root, and it worked when the output dir was normal and when I set the output to `/tmp` for testing, so I think it works as we want it to.

Reply:
Found a fun problem with it that I've been working to resolve, `compilation.emitAsset` doesn't accept an absolute path. Got it working by generating a relative path from the compilation output path to the context path via the `path.relative` method. Going to push a commit shortly.